### ISSUE TITLE
UnicodeEncodeError: 'charmap' codec can't encode character '\u2026' in position 997: character maps to <undefined>

### ISSUE THREAD

**Issue by ohYesOhNo** [dbcli/pgcli#628]
Executing this SQL: 

select * from trunk.str where str_name = 'abc'

Gives me this callstack (seems like an encoding issue):

Traceback (most recent call last):
  File "C:\Programs\Python\Scripts\pgcli-script.py", line 9, in <module>
    load_entry_point('pgcli==1.1.0', 'console_scripts', 'pgcli')()
  File "c:\programs\python\lib\site-packages\click\core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "c:\programs\python\lib\site-packages\click\core.py", line 696, in main
    rv = self.invoke(ctx)
  File "c:\programs\python\lib\site-packages\click\core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\programs\python\lib\site-packages\click\core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "c:\programs\python\lib\site-packages\pgcli\main.py", line 745, in cli
    pgcli.run_cli()
  File "c:\programs\python\lib\site-packages\pgcli\main.py", line 417, in run_cli
    click.echo_via_pager('\n'.join(output))
  File "c:\programs\python\lib\site-packages\click\termui.py", line 213, in echo_via_pager
    return pager(text + '\n', color)
  File "c:\programs\python\lib\site-packages\click\_termui_impl.py", line 287, in pager
    return _tempfilepager(text, 'more <', color)
  File "c:\programs\python\lib\site-packages\click\_termui_impl.py", line 357, in _tempfilepager
    f.write(text.encode(encoding))
  File "c:\programs\python\lib\encodings\cp852.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_map)
UnicodeEncodeError: 'charmap' codec can't encode character '\u2026' in position 997: character maps to <undefined>

**Comment by j-bennet** [comment]
@darikg Looks like a Windows issue to me. Would you be able to take a look (I don't run Windows)?

**Comment by darikg** [comment]
hi @ohYesOhNo what version of windows are you using? 32 or 64 bit? What console, cmd.exe or something else? And how did you install pgcli? 

If you're using cmd.exe, it might be worth trying instead something like [cmder](http://cmder.net/)

**Comment by ohYesOhNo** [comment]
This is on Windows 10 64bit; I see the same error with either ConEmu (https://conemu.github.io/) or with vanilla cmd; the Python version is 3.5.2, and I think I installed pgcli through 'pip install pgcli'.

**Comment by darikg** [comment]
I don't have Windows 10 handy to test this, but on Windows 7 I've had problems with the default system pager `more.exe` which apparently tries to guess text encodings. Do you have a different pager handy? I have `less.exe` which came with Git for windows. For me that looks like this:

```
temp> \pager less
PAGER set to less.
temp> select generate_series(1,10), round(random());
+-------------------+---------+
|   generate_series |   round |
|-------------------+---------|
|                 1 |       1 |
|                 2 |       1 |
|                 3 |       0 |
|                 4 |       0 |
|                 5 |       0 |
|                 6 |       0 |
|                 7 |       1 |
|                 8 |       1 |
|                 9 |       1 |
|                10 |       0 |
+-------------------+---------+
SELECT 10
temp> \pager more
PAGER set to more.
temp> select generate_series(1,10), round(random());
?????????????????+?????????+?????????????????????++++++++??+++%??++++++++'?+++%??++++++++??+++%??++++++++??+++???++++++++`?+++???++++++++??+++???++++++++??+++???++++++++??+++???++++++++<?+++???+++++++?%?+++%??????????????????????
```

You can see that the `more` pager makes a bizarre encoding error. I wonder a similar thing is happening to you.

**Comment by ohYesOhNo** [comment]
Here's what I did instead, because '\pager' doesn't do anything on my system:

1/ Created test.txt with '… … …' (these are the \u2026 characters that happened in the error)

2/ cat test.txt | less – prints '… … …' – ok

3/ cat test.txt | more – prints 'ÔÇŽ ÔÇŽ ÔÇŽ' – weird

(I also have the less/more that came with Git for Windows)

**Comment by ohYesOhNo** [comment]
I updated my pgcli to 1.3.1 and now I can do '\pager less' and it confirms that 'PAGER set to less'. However, strangely, the original issue still persists.

**Comment by darikg** [comment]
Ok, thanks for trying that. I can reproduce the problem on my end now:

```
C:\Users\dg> python -c "import click; click.echo_via_pager('… … …')"

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\dg\Anaconda3\envs\pgcli3\lib\site-packages\click\termui.py", line 203, in echo_via_pager
    return pager(text + '\n', color)
  File "C:\Users\dg\Anaconda3\envs\pgcli3\lib\site-packages\click\_termui_impl.py", line 262, in pager
    return _tempfilepager(text, os.environ['PAGER'], color)
  File "C:\Users\dg\Anaconda3\envs\pgcli3\lib\site-packages\click\_termui_impl.py", line 337, in _tempfilepager
    f.write(text.encode(encoding))
  File "C:\Users\dg\Anaconda3\envs\pgcli3\lib\encodings\cp437.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_map)
UnicodeEncodeError: 'charmap' codec can't encode character '\u2026' in position 0: character maps to <undefined>
```

I'll try to poke around in click and see if I can reduce it further.

**Comment by darikg** [comment]
Actually it's got nothing to do with click at all:

```
python -c "print('…')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\dg\Anaconda3\envs\pgcli3\lib\encodings\cp437.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_map)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u2026' in position 0: character maps to <undefined>
```
Can you try this minimal example on your machine?

**Comment by ohYesOhNo** [comment]
This produces exactly the same output on my end as well.

**Comment by darikg** [comment]
Are you able to upgrade to Python 3.6? It fixed the issue for me. According to [this](http://stackoverflow.com/a/32176732), 3.6 changed how unicode is handled on windows. To install pgcli, I used the psycopg2 binaries from [here](http://www.stickpeople.com/projects/python/win-psycopg/). If it's not possible or preferable to upgrade, it looks like you can install [win-unicode-console](https://github.com/Drekin/win-unicode-console) and enable that globally in site-packages somehow.

**Comment by ohYesOhNo** [comment]
I upgraded to Python 3.6 and here's the status:

- `python -c "print('…')"` now correctly outputs '…'
- With '…' in test.txt, `cat test.txt | more` outputs '╗┐ÔÇŽ'
- With '…' in test.txt, `cat test.txt | less` correctly outputs '…'
- In pgcli with the default pager, `select '…'` outputs 'ÔÇŽ'
- In pgcli with `\pager less`, `select '…'` correctly outputs '…'

So the issue seem to be related to `more` rather than to `pgcli`. I switched my default pager in `pgcli`'s  `config` to `less` (which, conveniently, was already there and just needed to be uncommented).

Thanks a lot for guiding me through this.
